File: src/recommendationservice/logger.py
#!/usr/bin/python

# Copyright The OpenTelemetry Authors
# SPDX-License-Identifier: Apache-2.0

import logging
import sys
from pythonjsonlogger import jsonlogger
from opentelemetry import trace


class CustomJsonFormatter(jsonlogger.JsonFormatter):
    def add_fields(self, log_record, record, message_dict):
        super(CustomJsonFormatter, self).add_fields(log_record, record, message_dict)
        if not log_record.get('otelTraceID'):
            log_record['otelTraceID'] = trace.format_trace_id(trace.get_current_span().get_span_context().trace_id)
        if not log_record.get('otelSpanID'):
            log_record['otelSpanID'] = trace.format_span_id(trace.get_current_span().get_span_context().span_id)

def getJSONLogger(name):
    logger = logging.getLogger(name)
    handler = logging.StreamHandler(sys.stdout)
    formatter = CustomJsonFormatter('%(asctime)s %(levelname)s [%(name)s] [%(filename)s:%(lineno)d] [trace_id=%(otelTraceID)s span_id=%(otelSpanID)s] - %(message)s')
    handler.setFormatter(formatter)
    logger.addHandler(handler)
    logger.setLevel(logging.INFO)
    logger.propagate = False
    return logger

File: src/paymentservice/logger.js
// Copyright The OpenTelemetry Authors
// SPDX-License-Identifier: Apache-2.0
const {
	LoggerProvider,
	SimpleLogRecordProcessor,
	ConsoleLogRecordExporter,
} = require("@opentelemetry/sdk-logs");

const loggerProvider = new LoggerProvider();
loggerProvider.addLogRecordProcessor(
	new SimpleLogRecordProcessor(new ConsoleLogRecordExporter())
);

module.exports = loggerProvider.getLogger("paymentservice", "1.0.0");